device/transport/amqp/test/_amqp_test.js

```diff
@@ -4,246 +4,1345 @@
+
+      it('attaches the C2D link if a listener is registered for the \'message\' event', function (testCallback) {
+        transport.on('message', function () {});
+        transport.connect(function () {
+          assert.isTrue(fakeBaseClient.attachReceiverLink.calledWith(transport._c2dEndpoint));
+          testCallback();
+        });
+      });
+
+      it('subscribes to the \'message\' event on the C2D link if a listener is registered for the \'message\' event', function (testCallback) {
+        var fakeReceiverLink = new EventEmitter();
+        var fakeMessageHandler = sinon.spy();
+        var fakeMessage = new Message('foo');
+        transport.on('message', fakeMessageHandler);
+
+        transport.connect(function () {
+          receiver.emit('message', fakeMessage);
+          assert(fakeMessageHandler.calledWith(fakeMessage));
+          testCallback();
+        });
+      });
+
+      it('subscribes to the \'error\' event on the C2D link', function (testCallback) {
+        transport.connect(function () {
+          transport.on('message', function () {});
+          assert.isTrue(receiver.on.calledWith('error'));
+          testCallback();
+        })
       });
     });
 
-    it('invokes puttoken and an error results', function () {
-      var testError = new Error('fake error');
-      var transport = new Amqp(configWithSAS);
-      sinon.stub(transport._amqp, 'putToken').callsArgWith(2, testError);
-      transport.updateSharedAccessSignature(simpleSas, function(err) {
-        assert.instanceOf(err, Error);
-        assert.strictEqual(err.amqpError, testError);
+    describe('#disconnect', function () {
+      describe('if disconnected already:', function () {
+        /*Tests_SRS_NODE_DEVICE_AMQP_16_010: [The `done` callback method passed in argument shall be called when disconnected.]*/
+        it('calls the callback immediately if a callback is specified', function (testCallback) {
+          transport.disconnect(function (err, result) {
+            assert.isNull(err);
+            assert.instanceOf(result, results.Disconnected);
+            testCallback();
+          });
+        });
+
+        it('does not crash if no callback is specified', function () {
+          assert.doesNotThrow(function () {
+            transport.disconnect();
+          });
+        });
+      });
+
+      describe('if called while connecting:', function () {
+        it('is deferred until connecting fails', function (testCallback) {
+          var connectErr = new Error('cannot connect');
+          var connectCallback;
+          fakeBaseClient.connect = sinon.stub().callsFake(function (uri, options, done) {
+            connectCallback = done;
+          });
+
+          transport.connect(function (err) {
+            assert.strictEqual(err.amqpError, connectErr);
+            fakeBaseClient.connect = sinon.stub().callsArgWith(2, null, new results.Connected());
+          });
+
+          // now blocked in "connecting" state
```

Comment:
>// now blocked in "connecting" state [](start = 10, length = 36)

And you know this how?  And it's testable how?

Reply:
because I stubbed the connect method on line 443.  I'll update to a better comment. 

---
In reply to: [141126023](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141126023) [](ancestors = 141126023)